### PR TITLE
[#808] fix(client): use `synchronized` to make `spill` and `addRecord` thread-safe

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

1. Use the `synchronized` to make `spill` and `addRecord` thread safe.

### Why are the changes needed?

Fix: #808 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?
1. UTs

### CODE REVIEW

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 1]
```diff
@@ -204,26 +207,29 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
...
-    LOG.info("Flush total buffer for shuffleId[" + shuffleId + "] with allocated["
-        + allocatedBytes + "], dataSize[" + dataSize + "], memoryUsed[" + memoryUsed + "]");
-    buffers.clear();
     return result;
   }
 
   // transform records to shuffleBlock
   protected ShuffleBlockInfo createShuffleBlock(int partitionId, WriterBuffer wb) {
-    byte[] data = wb.getData();
+    byte[] data = wb.finalizeAndGetData();
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: Why do we need `finalizeAndGetData`?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 2]
```diff
@@ -204,26 +207,29 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
   }
 
   // transform all [partition, records] to [partition, ShuffleBlockInfo] and clear cache
-  public synchronized List<ShuffleBlockInfo> clear() {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:207]: It's more clear to make `clear` and `addRecord` synchronized.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java (hunk 1 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: This is a defensive programming. Once the data is gotten, the buffer is aborted.

Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: > This is a defensive programming. Once the data is gotten, the buffer is aborted.

Is it necessary? A little weird.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: I think yes. without this, the other developer may addRecord again after finalize. 

BTW, the similar implementation scatter everywhere

Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: Could you paste some example code?

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java like this.

Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: > https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java like this.

It don't have method like `finalizeAndGetData`.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:232]: Let me remove this part

**Review by jerqi — DISMISSED** [review]
LGTM. You would better to verify this patch with your production job.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 3]
```diff
@@ -130,7 +131,7 @@ public WriteBufferManager(
     this.bufferSize = bufferManagerOptions.getBufferSize();
     this.spillSize = bufferManagerOptions.getBufferSpillThreshold();
     this.instance = serializer.newInstance();
-    this.buffers = Maps.newHashMap();
+    this.buffers = Maps.newConcurrentMap();
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:134]: Yes. Not necessary.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 4]
```diff
@@ -173,25 +174,27 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
       return null;
     }
     List<ShuffleBlockInfo> result = Lists.newArrayList();
-    if (buffers.containsKey(partitionId)) {
-      WriterBuffer wb = buffers.get(partitionId);
-      if (wb.askForMemory(serializedDataLength)) {
+    synchronized (buffers) {
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:177]: I'm exploring. let me try.

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:177]: The key of keeping WriteBuffer thread-safe is about `addRecord` and `getData`. I have not better ways.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 5]
```diff
@@ -76,7 +77,7 @@ public class WriteBufferManager extends MemoryConsumer {
   private Map<Integer, List<ShuffleServerInfo>> partitionToServers;
   private int serializerBufferSize;
   private int bufferSegmentSize;
-  private long copyTime = 0;
+  private volatile long copyTime = 0;
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java]: Done

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java [hunk 6]
```diff
@@ -26,70 +29,70 @@
...
-  private int nextOffset = 0;
+  private AtomicLong copyTime = new AtomicLong(0);
+  private AtomicReference<byte[]> buffer = new AtomicReference<>();
+  private final AtomicInteger bufferSize;
+  private AtomicInteger nextOffset = new AtomicInteger(0);
   private List<WrappedBuffer> buffers = Lists.newArrayList();
-  private int dataLength = 0;
-  private int memoryUsed = 0;
+  private AtomicInteger dataLength = new AtomicInteger();
+  private AtomicInteger memoryUsed = new AtomicInteger();
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: Absolutely not. The `addRecord` and `getData` will be accessed by 2 threads.

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: Once `getData` is invoked, the `writeBuffer` should be changed to read-only. 

This `semantic` is not shown in this part.

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: Oh, I forget this. Let's revert.

[the remark this answers is not in the document]
Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: Atomic type is not needed, but the volatile is needed, isn't it?

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: I think yes. The double check mechanism is like this.

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java:38]: I think yes. The double check mechanism is like this.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java]: If the operation of `copyTime` will be operated in the scope of lock, they don't nee `volatile`, too.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriterBuffer.java]: Recheck this part, I think we don't need `volatile`

File: client-spark/common/src/test/java/org/apache/spark/shuffle/writer/WriteBufferManagerTest.java [hunk 7]
```diff
@@ -221,6 +227,58 @@ public void buildBlockEventsTest() {
...
+    }).start();
+
+
+    new Thread(() -> {
+      while (!produceFinished.get()) {
+        spyManager.spill(100000, mock(WriteBufferManager.class));
+      }
+      spyManager.spill(100000, mock(WriteBufferManager.class));
+      spillFinished.set(true);
+    }).start();
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/test/java/org/apache/spark/shuffle/writer/WriteBufferManagerTest.java:276]: Yes. But it's guarded by `TaskMemoryManager` lock

a1

[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/test/java/org/apache/spark/shuffle/writer/WriteBufferManagerTest.java]: Done

**Comment by zuston** [comment]
After rethinking, I think the spill should not block when acquiring the lock, because the addRecord can take a long time, especially for tight memory.

So in the latest commit. I use the `ReentrantLock#tryLock` to avoid this.

**Comment by zuston** [comment]
> Do you have any cases report that improves for the new fix? Seems that it would be quite possible that executor with memory pressure and high write frequency, the spill could not take effect due to that it cannot acquire the lock, which would still causes executor OOM.

When cherry-pick this PR to internal uniffle, I found this problem but I don't validate.

> Do you think is it possible to adaptive increate try lock wait time to alleviate that cases?

I think yes. We'd better to avoid acquiring memory blocked.

**Comment by zuston** [comment]
I think I will freeze this until all are fixed

**Comment by zuston** [comment]
I will close this PR. Better solution is #848  cc @advancedxy

**Pull request closed without merging**